internal/vm/vm.go

```diff
@@ -0,0 +1,277 @@
+	ErrNotInRunningState    = errors.New("VM is not in running state")
+	ErrNotInPausedState     = errors.New("VM is not in paused state")
+)
+
+// UVMSource is an object that can launch a Windows or Linux Utility VM.
+type UVMSource interface {
+	NewLinuxUVM(ctx context.Context, id string, owner string) (UVM, error)
+	NewWindowsUVM(ctx context.Context, id string, owner string) (UVM, error)
+}
+
+const (
+	HCS      = "hcs"
+	RemoteVM = "remotevm"
+)
+
+// UVM is an abstraction around a lightweight virtual machine. It houses core lifecycle methods such as Create
+// Start, and Stop and also several optional nested interfaces that can be used to determine what the virtual machine
+// supports and to configure these resources.
+type UVM interface {
+	// ID will return a string identifier for the Utility VM.
+	ID() string
+
+	// State returns the current running state of the Utility VM. e.g. Created, Running, Terminated
+	State() State
+
+	// Create will create the Utility VM in a paused/powered off state with whatever is present in the implementation
+	// of the interfaces config at the time of the call.
+	Create(ctx context.Context) error
+
+	// Start will power on the Utility VM and put it into a running state. This will boot the guest OS and start all of the
+	// devices configured on the machine.
+	Start(ctx context.Context) error
+
+	// Stop will shutdown the Utility VM and place it into a terminated state.
+	Stop(ctx context.Context) error
+
+	// Pause will place the Utility VM into a paused state. The guest OS will be halted and any devices will have be in a
+	// a suspended state. Save can be used to snapshot the current state of the virtual machine, and Resume can be used to
+	// place the virtual machine back into a running state.
+	Pause(ctx context.Context) error
+
+	// Resume will put a previously paused Utility VM back into a running state. The guest OS will resume operation from the point
+	// in time it was paused and all devices should be un-suspended.
+	Resume(ctx context.Context) error
+
+	// Save will snapshot the state of the Utility VM at the point in time when the VM was paused.
+	Save(ctx context.Context) error
+
+	// Wait synchronously waits for the Utility VM to shutdown or terminate. A call to stop will trigger this
+	// to unblock.
+	Wait() error
+
+	// Stats returns statistics about the Utility VM. This includes things like assigned memory, available memory,
+	// processor runtime etc.
+	Stats(ctx context.Context) (*stats.VirtualMachineStatistics, error)
+
+	// ExitError will return any error if the Utility VM exited unexpectedly, or if the Utility VM experienced an
+	// error after Wait returned, it will return the wait error.
+	ExitError() error
+
+	MemoryManager
+	ProcessorManager
+	StorageQosManager
+	SerialManager
+	BootManager
+	SCSIManager
+	VSMBManager
+	Plan9Manager
+	VPMemManager
+	NetworkManager
+	PCIManager
+	VMSocketManager
+	WindowsConfigManager
+	LinuxConfigManager
```

Comment:
This is what I'd like the most feedback on. Currently everything HCS supports is filled out, but if we're going through a different virtstack we don't know what is. The idea was to have a call we can make to determine what exactly is supported and then use this information to determine what to do when we'd make a call like `AddSCSI` or similar. 

Can imagine a quick check at the top of every call like 
```go
func (uvm *utilityVM) AddSCSIDisk(ctx context.Context, controller uint32, lun uint32, path string, typ vm.SCSIDiskType, readOnly bool) error {
      if !uvm.supported.SCSI {
             return vm.ErrNotSupported
      }
      // Rest of the function...
}
```

This makes it so when calling something we have a known error we can check against (`vm.ErrNotSupported`) to determine if we should try something else as it's not supported or if it's absolutely necessary we'd just bubble it up and fail. Good example would be if SCSI was supported but VPMem wasn't, during LCOW layer adds we could check if `AddVPMem` returned  `vm.ErrNotSupported` and immediately just skip any further calls and go try SCSI.

This does make things pretty drop in, and the only extra boilerplate is checking this new error (and only in cases where we can do something useful with it). Although I'm wondering whether it might be better to have `IsXSupported` methods on each interface so we can know upfront. 

@kevpar had suggested another approach in a similar vein of the `IsXSupported` idea of exposing methods like this on the main UVM interface instead.
```go 
SCSI() SCSIManager
VPMem() VPMemManager
```
and instead you'd call these and either get a nil value if <thing> isn't supported or the interfaces that are currently embedded into `UVM` at the moment if it is. This does add a whole bunch of 
```go
scsi := uvm.u.SCSI()
if scsi == nil {
     return errors.New("oh no)
} 
```
's everywhere but it might be cleaner. I don't have a strong opinion here.

Reply:
Some variant of IsxSupported seems preferable as it should be easier to code to the interface when what is supported can be determined up front. For instance relying on AddSCSIDisk to fail for discovery forces the caller to potentially unwind and take a different path.

Reply:
The reasoning right now is we don't have many things at all that we'd go a different route if it failed or not. Ideally if the client asked for a pci device to be added and the virtstack didn't support it, we just hard fail with no fallback. Same for if we can't mount in a directory they asked for because VSMB's not operational etc. The only one really is during layer adds if VPMem failed for some specific reasons we just move on and try adding the same disk over scsi, which isn't very hard to achieve with what's here as we just add in one extra known error that will let us continue if received. 

The current approach keeps the boilerplate to a minimum but it is a bit unintuitive to call something and check the error to determine if it's supported, rather than checking upfront (even if it's not needed). I'm fine with going with adding `IsXSupported` to everything and calling that before all of the resource operations, or if everyone likes Kevin's idea also we can go that route.

Reply:
Presumably the caller can just skip the IsSupported checks for cases where there is no fallback so there shouldn't be much extra boiler plate code.

Reply:
That's a good point! Ideally whatever we're talking to would return some sort of "isn't supported" error also anyways if the virtstack doesn't support it, so at least the error strings still informational on the ones with no fallback.

Reply:
Might be better to just add one `Supported` call that takes in a constant of what resource we're asking about instead of a call on every interface. Might also need a second parameter in case adding is supported but remove isn't or adding is but updating isn't etc. 

e.g.
```go
type Resource uint8

const (
    SCSI = iota
    PCI 
   // etc.
)

// And then caller would invoke something like the below
if vm.Supported(vm.PCI, vm.HotAdd) {
   err := uvm.u.AddPCIDevice(ctx, instanceID)
} else {
    // Abort mission
}

Reply:
I think it's going to be a pain to add more methods to the interface with this approach. I think you want something that keeps the `UVM` interface itself short.

Reply:
I think that means both an interface (to keep the root interface small) and a supported check (to allow runtime opt-in). So you might need something like:

```
vpmem, ok := uvm.u.(VPMEM)
if !ok || !scsi.VPMEMSupported() {
    /// fall back to SCSI
}
```

Reply:
Yea.. Even with the "call SCSI() and get back a SCSIManager" approach at least the methods are off of UVM itself and self contained in their own interfaces.

Reply:
I am kind of fond of the 
```go
scsi := uvm.u.SCSI()
if scsi == nil {
     // abort mission 
}
// use SCSIManager
```
approach instead of trying to typecast, but I haven't actually given this a go.

Reply:
If we keep adding functionality that's specific to one virt stack or another, it will be really annoying to have to update each UVM implementation to add boilerplate `return nil` or whatever (times two for the builder case). That's why having a separate interface for the optional stuff will be useful.

Reply:
Are you referring to typecasting might be a better move to avoid filling out the boilerplate 
```go
func (uvm *utilityVM5) PCI() PCIManager {
      // Assume capabilities of the virtstack have been 
     // determined already.
     if uvm.capabilities.pci {
           return &pci{}
     }
     return nil 
}
```
for every implementation? And then on a 
```go
uvm.u.Supported(vm.VPMEM, vm.HotAdd)
```
call we'd do the capabilities check inside instead?